rust/src/composepost.rs

```diff
@@ -676,30 +676,33 @@ fn convert_path_to_tmpfiles_d_recurse(
             continue;
         }
 
-        let filetype_char = match path_type {
-            SimpleType::Dir => 'd',
-            SimpleType::Symlink => 'L',
-            SimpleType::File => 'f',
-            x => unreachable!("invalid path type: {:?}", x),
-        };
-        write!(tmpfiles_bufwr, "{} ", filetype_char)?;
-        write!(tmpfiles_bufwr, "/{} ", full_path)?;
-
-        if path_type == SimpleType::Symlink {
-            let link_target = rootfs.read_link(&full_path)?;
-            write!(tmpfiles_bufwr, "- - - - ")?;
-            write!(tmpfiles_bufwr, "{}", link_target.display())?;
-        } else {
+        // Translate this file entry.
+        let entry = {
             let meta = rootfs.metadata(&full_path)?;
-            let perm = meta.stat().st_mode & !libc::S_IFMT;
-            write!(tmpfiles_bufwr, "{:04o} ", perm)?;
+            let mode = meta.stat().st_mode & !libc::S_IFMT;
+
+            let file_info = gio::FileInfo::new();
+            file_info.set_attribute_uint32("unix::mode", mode);
+
+            match path_type {
+                SimpleType::Dir => file_info.set_file_type(FileType::Directory),
+                SimpleType::Symlink => {
+                    file_info.set_file_type(FileType::SymbolicLink);
+                    let link_target = rootfs.read_link(&full_path)?;
+                    file_info.set_symlink_target(
+                        link_target.to_str().expect("non UTF-8 symlink target"),
```

Comment:
Could use `camino::Utf8Path` to propagate the error and also ensure the error text includes the path.